### PR TITLE
Topo sort the model before saving

### PR DESCRIPTION
**Pull request by wangyems**
**Description**: Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by jywu-msft** [comment]
why is there also onnx_model.py (with topological_sort code) in python/tools/quantization as well?
Is it possible to share the same code between the modules?

**Review by tianleiwu — APPROVED** [review]

**Comment by tianleiwu** [comment]
It might be better to call it in optimize_model since model is quantized like quantize_model(optimize_model()) and finally saved to model file. That means the input for quantization might not be topo sorted.

File: onnxruntime/python/tools/transformers/onnx_model.py [hunk 1]
```diff
@@ -781,7 +781,65 @@ def is_safe_to_fuse_nodes(self, nodes_to_remove, keep_outputs, input_name_to_nod
                             return False
         return True
 
+    def graph_topological_sort(self, graph):
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model.py:784]: This function could be a static method and called in the topological_sort function of quantization.

File: onnxruntime/python/tools/transformers/onnx_model.py [hunk 2]
```diff
@@ -781,7 +781,65 @@ def is_safe_to_fuse_nodes(self, nodes_to_remove, keep_outputs, input_name_to_nod
...
+                if output in deps_to_nodes:
+                    for node_idx in deps_to_nodes[output]:
+                        deps_count[node_idx] = deps_count[node_idx] - 1
+                        if deps_count[node_idx] == 0:
+                            sorted_nodes.append(graph.node[node_idx])
+                            end = end + 1
+            start = start + 1
+
+        # TODO: This assertion does not apply to the subgraphs, need other type of check
+        #assert(end == len(graph.node)), "Graph is not a DAG"
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model.py:831]: If end != len(graph.node), that means some node is in graph.node but not in sorted_nodes output. Then sounds something is wrong.

File: onnxruntime/python/tools/transformers/onnx_model.py [hunk 3]
```diff
@@ -781,7 +781,65 @@ def is_safe_to_fuse_nodes(self, nodes_to_remove, keep_outputs, input_name_to_nod
...
+                sorted_nodes.append(graph.node[node_idx])
+                continue
+
+            for input_name in node.input:
+                if input_name not in deps_to_nodes:
+                    deps_to_nodes[input_name] = [node_idx]
+                else:
+                    deps_to_nodes[input_name].append(node_idx)
+
+        initializer_names = [init.name for init in graph.initializer]
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model.py:801]: For subgraph, initializers can be from subgraph and also from parent graph. So this logic only applies to top level graph.

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**